### PR TITLE
Use containers based on their manifest

### PR DESCRIPTION
**Pull request by carabasdaniel**

### CODE REVIEW

File: pkg/app/images.go [hunk 1]
```diff
@@ -52,13 +59,22 @@ func (c *PolicyApp) Images() error {
...
+				manifest.Annotations[ocispec.AnnotationCreated],
+				strings.ReplaceAll(humanize.Bytes(uint64(descr.Size)), " ", "")}
+		} else {
+			arrData = []string{
+				familiarName,
+				tagOrNone,
+				descr.Digest.Encoded()[:12],
+				descr.Annotations[ocispec.AnnotationCreated],
+				strings.ReplaceAll(humanize.Bytes(uint64(descr.Size)), " ", "")}
+		}
```
Comment (BogdanIrimie) [inline pkg/app/images.go:77]: I would inverse the if and else to make it easier to read (the else is a double negation not not equal to nil => equal to nil)

```go
		if manifest == nil {
			arrData = []string{
				familiarName,
				tagOrNone,
				descr.Digest.Encoded()[:12],
				descr.Annotations[ocispec.AnnotationCreated],
				strings.ReplaceAll(humanize.Bytes(uint64(descr.Size)), " ", "")}
		} else {
			arrData = []string{
				familiarName,
				tagOrNone,
				descr.Digest.Encoded()[:12],
				manifest.Annotations[ocispec.AnnotationCreated],
				strings.ReplaceAll(humanize.Bytes(uint64(descr.Size)), " ", "")}
		}
```

Comment (BogdanIrimie) [reply pkg/app/images.go:77]: Come to thing of it I would even go a step further and refactor it to
```go
		var annotations string
		if manifest == nil {
			annotations = descr.Annotations[ocispec.AnnotationCreated]
		} else {
			annotations = manifest.Annotations[ocispec.AnnotationCreated]
		}
		
		arrData := []string{
			familiarName,
			tagOrNone,
			descr.Digest.Encoded()[:12],
			annotations,
			strings.ReplaceAll(humanize.Bytes(uint64(descr.Size)), " ", "")}
```

File: pkg/app/inspect.go [hunk 2]
```diff
@@ -23,14 +24,27 @@ func (c *PolicyApp) Inspect(userRef string) error {
...
+		if len(manifest.Annotations) > 0 {
+			msg := c.UI.Normal().WithTable("Annotation", "Value")
+
+			for k, v := range manifest.Annotations {
+				msg.WithTableRow(k, v)
+			}
+			msg.Msg("Annotations")
+		}
+
+	} else if len(contentInfo.Annotations) > 0 {
```
Comment (BogdanIrimie) [inline pkg/app/inspect.go:47]: I would refactor this and extract the annotation logic in a method and have only one for that cycles through annotations
```go
package app

import (
	"github.com/opcr-io/policy/pkg/oci"
	"github.com/opcr-io/policy/pkg/parser"
	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
	"github.com/pkg/errors"
)

func (c *PolicyApp) Inspect(userRef string) error {
	defer c.Cancel()

	ref, err := parser.CalculatePolicyRef(userRef, c.Configuration.DefaultDomain)
	if err != nil {
		return err
	}

	ociClient, err := oci.NewOCI(c.Context, c.Logger, c.getHosts, c.Configuration.PoliciesRoot())
	if err != nil {
		return err
	}

	contentInfo, err := ociClient.GetStore().Resolve(c.Context, ref)
	if err != nil {
		return errors.Wrapf(err, "failed to read content info for policy [%s]", ref)
	}
	c.UI.Normal().
		WithStringValue("media type", contentInfo.MediaType).
		WithStringValue("digest", contentInfo.Digest.String()).
		WithIntValue("size", contentInfo.Size).
		Do()

	annotations, err := getAnnotations(contentInfo, ociClient)
	if err != nil {
		return err
	}

	msg := c.UI.Normal().WithTable("Annotation", "Value")

	for k, v := range annotations {
		msg.WithTableRow(k, v)
	}
	msg.Msg("Annotations")

	return nil
}

func getAnnotations(contentInfo ocispec.Descriptor, ociClient *oci.Oci) (map[string]string, error) {
	var annotations map[string]string
	if contentInfo.MediaType == ocispec.MediaTypeImageManifest {
		manifest, err := ociClient.GetManifest(&contentInfo)
		if err != nil {
			return nil, err
		}
		annotations = manifest.Annotations
	} else if len(contentInfo.Annotations) > 0 {
		annotations = contentInfo.Annotations
	}

	return annotations, nil
}
```

File: pkg/app/repl.go [hunk 3]
```diff
@@ -31,6 +32,20 @@ func (c *PolicyApp) Repl(ref string, maxErrors int) error {
...
+	if descriptor.MediaType == ocispec.MediaTypeImageManifest {
+		bundleDescriptor, _, err := ociClient.GetTarballAndConfigLayerDescriptor(c.Context, &descriptor)
+		if err != nil {
+			return err
+		}
+		bundleHex = bundleDescriptor.Digest.Hex()
+		if bundleHex == "" {
+			return errors.New("current manifest does not contain a MediaTypeImageLayerGzip")
+		}
+	}
```
Comment (BogdanIrimie) [inline pkg/app/repl.go:47]: I would extract this in a new method e.g.
```go
func getBundleHex(descriptor ocispec.Descriptor, ociClient *oci.Oci) (string, error) {
	var bundleHex string
	// check for media type - if manifest get tarbarll digest hex.
	if descriptor.MediaType == ocispec.MediaTypeImageManifest {
		bundleDescriptor, _, err := ociClient.GetTarballAndConfigLayerDescriptor(c.Context, &descriptor)
		if err != nil {
			return "", err
		}
		bundleHex = bundleDescriptor.Digest.Hex()
		if bundleHex == "" {
			return "", errors.New("current manifest does not contain a MediaTypeImageLayerGzip")
		}
	} else {
		bundleHex = descriptor.Digest.Hex()
	}

	return bundleHex, nil
}
```

File: pkg/app/rm.go [hunk 4]
```diff
@@ -96,48 +96,71 @@ func (c *PolicyApp) Rm(existingRef string, force bool) error {
...
+	indexBytes, err := os.ReadFile(indexPath)
+	if err != nil {
+		return false, err
+	}
+
+	err = json.Unmarshal(indexBytes, &localIndex)
+	if err != nil {
+		return false, err
+	}
+	for i := range localIndex.Manifests {
```
Comment (BogdanIrimie) [inline pkg/app/rm.go:147]: Could be replaced with
```go
	for _, descriptor := range localIndex.Manifests {
```

File: pkg/app/save.go [hunk 5]
```diff
@@ -35,6 +35,15 @@ func (c *PolicyApp) Save(userRef, outputFilePath string) error {
...
 	}
 
+	// if the reference descriptor is the manifest get the tarball descriptor information from the manifest layers.
+	if refDescriptor.MediaType == ocispec.MediaTypeImageManifest {
+		bundleDescriptor, _, err := ociClient.GetTarballAndConfigLayerDescriptor(c.Context, &refDescriptor)
+		if err != nil {
+			return err
+		}
+		refDescriptor = *bundleDescriptor
+	}
```
Comment (BogdanIrimie) [inline pkg/app/save.go:45]: Maybe extract this in a method

```go
func (c *PolicyApp)getRefDescriptor(ociClient *oci.Oci, userRef string) (*ocispec.Descriptor, error) {
	ref, err := parser.CalculatePolicyRef(userRef, c.Configuration.DefaultDomain)
	if err != nil {
		return nil, err
	}

	refs, err := ociClient.ListReferences()
	if err != nil {
		return nil, err
	}

	refDescriptor, ok := refs[ref]
	if !ok {
		return nil, errors.Errorf("policy [%s] not found in the local store", ref)
	}

	// if the reference descriptor is the manifest get the tarball descriptor information from the manifest layers.
	if refDescriptor.MediaType == ocispec.MediaTypeImageManifest {
		bundleDescriptor, _, err := ociClient.GetTarballAndConfigLayerDescriptor(c.Context, &refDescriptor)
		if err != nil {
			return nil, err
		}
		refDescriptor = *bundleDescriptor
	}
	
	return &refDescriptor, nil
}
```

File: pkg/app/rm.go (hunk 4 above)
Comment (carabasdaniel) [reply pkg/app/rm.go:147]: Avoid  G601: Implicit memory aliasing in for loop. (gosec)

**Review by BogdanIrimie — APPROVED** [review]

**Review by florindragos — APPROVED** [review]

**Pull request merged**